### PR TITLE
Staging

### PR DESCRIPTION
**Pull request by vamc-k**
I designed a rest API's for the Reminder blog.In this I have implemented
ModelSerializers with Validations
Token based authentication.
-Only authorized user can create Reminder and can fetch/update/delete only the Reminders he/she posted.
can you review the code @rhlk92 .

### CODE REVIEW

File: rest_blog/views.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+@csrf_exempt
+def reminder_list(request):
+    """
+    List all Reminders, or create a new Reminder.
+    """
+    if request.method == 'GET':
+        reminders = Reminder.objects.all()
+        serializer = ReminderSerializer(reminders, many=True)
+        return JsonResponse(serializer.data, safe=False)
```
Comment (rahulkarkera) [inline rest_blog/views.py]: Instead of using `JSONResponse`, use `Response` from django rest framework.

File: rest_blog/views.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    except Reminder.DoesNotExist:
+        return HttpResponse(status=404)
+
+    if request.method == 'GET':
+        serializer = ReminderSerializer(reminder)
+        return JsonResponse(serializer.data)
+
+    elif request.method == 'PUT':
+        data = JSONParser().parse(request)
+        serializer = ReminderSerializer(reminder, data=data)
```
Comment (rahulkarkera) [inline rest_blog/views.py]: @vamsikrishna-kolli You don't have to use `JSONParser `. Just pass `request.data` to the `ReminderSerializer`.

File: rest_blog/urls.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from django.conf.urls import url
+from rest_blog import views
+
+urlpatterns = [
+    url(r'^reminders/$', views.reminder_list),
```
Comment (rahulkarkera) [inline rest_blog/urls.py]: @vamsikrishna-kolli URL looks good.

File: rest_blog/serializers.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+from rest_framework import serializers
+from blog.models import Reminder
+
+
+class ReminderSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Reminder
+        fields = ('id', 'title', 'message', 'reminder_time', 'created_time', 'modified_time')
+
+    def validate_reminder_time(self, reminder_time):
```
Comment (rahulkarkera) [inline rest_blog/serializers.py]: @vamsikrishna-kolli Good use of validate method. Always remember not to commit print statements.

File: Reminder/urls.py [hunk 5]
```diff
@@ -13,12 +13,33 @@
...
+
+# ViewSets define the view behavior.
+class UserViewSet(viewsets.ModelViewSet):
+    queryset = User.objects.all()
+    serializer_class = UserSerializer
+
+
+# Routers provide an easy way of automatically determining the URL conf.
+router = routers.DefaultRouter()
+router.register(r'users', UserViewSet)
```
Comment (rahulkarkera) [inline Reminder/urls.py]: @vamsikrishna-kolli Move the `UserSerializer` to `serializers.py` and  `UserViewSet` to `views.py`.

File: blog/models.py [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
+
 
 class Reminder(models.Model):
-    id = models.AutoField
     title = models.CharField(max_length=50)
     message = models.CharField(max_length=50)
-    reminderTime = models.DateTimeField()
-    createdTime = models.DateTimeField(default=timezone.now())
-    modifiedTime = models.DateTimeField(default=timezone.now())
+    reminder_time = models.DateTimeField()
```
Comment (rahulkarkera) [inline blog/models.py:11]: @vamsikrishna-kolli You made all the case changes. Good.

File: Reminder/settings.py [hunk 7]
```diff
@@ -37,9 +37,19 @@
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
-    'blog',
+    'rest_framework',
+    'blog.apps.BlogConfig',
+    'rest_blog.apps.RestBlogConfig',
```
Comment (rahulkarkera) [inline Reminder/settings.py]: @vamsikrishna-kolli This is good. You can do this also. But in this situation, you don't have any particular Application configurations. So you can remove the `apps.py` file. And then just put it as `blog` and  `rest_blog`.

File: rest_blog/views.py (hunk 1 above)
Comment (vamc-k) [reply rest_blog/views.py]: I have updated Response in the Class based views.
I'll update the code.

File: Reminder/urls.py (hunk 5 above)
Comment (rahulkarkera) [reply Reminder/urls.py]: @vamsikrishna-kolli Remove `UserSerializer` and `UserViewSet`.

File: rest_blog/views_class_based.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        return Reminder.objects.all()
+
+    def get_object(self, pk):
+        reminders = self.get_queryset()
+        try:
+            reminder = reminders.get(pk=pk)
+            self.check_object_permissions(self.request, reminder)
+        except Reminder.DoesNotExist:
+            reminder = 204
+        return reminder
```
Comment (rahulkarkera) [inline rest_blog/views_class_based.py]: @vamsikrishna-kolli Move the check object permissions somewhere else. Here just check for the object and raise exception if object does not exist. If you raise the exception here, you don't have to check again `if reminder == 204`. Checkout exceptions for this. Make the changes everywhere.
And also go through different error codes and see what they mean.

File: rest_blog/views_class_based.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    def delete(self, request, pk, format=None):
+        reminder = self.get_object(pk)
+        if reminder == 204:
+            content = {
+                'status': 204,
+                'message': 'No reminder with the id you provided.'
+            }
+            return Response(content)
+        reminder.delete()
+        content = {
```
Comment (rahulkarkera) [inline rest_blog/views_class_based.py]: Check what error codes they use when you delete an object.

File: rest_blog/views.py (hunk 1 above)
Comment (rahulkarkera) [reply rest_blog/views.py]: Remove this view. Use only the class based view.

**Review by rahulkarkera — APPROVED** [review]
LGTM

**Pull request merged**